runtime/gpufortrt/src/gpufortrt_api.f90

```diff
@@ -0,0 +1,850 @@
+      end subroutine
+    end interface 
+    !
+    if ( present(async_arg) ) then
+      call gpufortrt_data_end_async_c_impl(async_arg)  
+    else
+      call gpufortrt_data_end_c_impl()  
+    endif
+  end subroutine
+
+  subroutine gpufortrt_enter_exit_data(mappings,async_arg,finalize)
+    use iso_c_binding
+    implicit none
+    !
+    !integer,intent(in) :: device_kind
+    type(gpufortrt_mapping_t),dimension(:),target,intent(in),optional :: mappings
+    integer(gpufortrt_handle_kind),intent(in),optional :: async_arg
+    logical,intent(in),optional :: finalize
+    !
+    interface
+      subroutine gpufortrt_enter_exit_data_c_impl(mappings,num_mappings,finalize) bind(c,name="gpufortrt_enter_exit_data")
+        use iso_c_binding
+        implicit none
+        type(c_ptr),value,intent(in) :: mappings
+        integer(c_int),value,intent(in) :: num_mappings
+        logical(c_bool),value,intent(in) :: finalize
+      end subroutine
+      subroutine gpufortrt_enter_exit_data_async_c_impl(mappings,num_mappings,async_arg,finalize) bind(c,name="gpufortrt_enter_exit_data_async")
+        use iso_c_binding
+        use gpufortrt_types, only: gpufortrt_handle_kind
+        implicit none
+        type(c_ptr),value,intent(in) :: mappings
+        integer(c_int),value,intent(in) :: num_mappings
+        integer(gpufortrt_handle_kind),value,intent(in) :: async_arg
+        logical(c_bool),value,intent(in) :: finalize
+      end subroutine
+    end interface 
+    !
+    logical(c_bool) :: opt_finalize
+    !
+    opt_finalize = .false._c_bool
+    if ( present(finalize) ) opt_finalize = logical(finalize,kind=c_bool)
+    !
+    if ( present(async_arg) ) then
+      if ( present(mappings) ) then
+        call gpufortrt_enter_exit_data_async_c_impl(&
+          c_loc(mappings),&
+          size(mappings,kind=c_int),&
+          async_arg,&
+          opt_finalize) 
+      else
+        call gpufortrt_enter_exit_data_async_c_impl(&
+          c_null_ptr,&
+          0_c_int,&
+          async_arg,&
+          opt_finalize) 
+      endif
+    else
+      if ( present(mappings) ) then
+        call gpufortrt_enter_exit_data_c_impl(&
+          c_loc(mappings),&
+          size(mappings,kind=c_int),&
+          opt_finalize) 
+      else
+        call gpufortrt_enter_exit_data_c_impl(&
+          c_null_ptr,&
+          0_c_int,&
+          opt_finalize) 
+      endif
+    endif
+  end subroutine
+
+  !> Lookup device pointer for given host pointer.
+  !> \param[in] condition condition that must be met, otherwise host pointer is returned. Defaults to '.true.'.
+  !> \param[in] if_present Only return device pointer if one could be found for the host pointer.
+  !>                       otherwise host pointer is returned. Defaults to '.false.'.
+  !> \note Returns a c_null_ptr if the host pointer is invalid, i.e. not C associated.
+  subroutine gpufortrt_use_device(resultptr, hostptr,sizes,lbounds,if_arg,if_present) 
```

Comment:
I'm working on use_device, as it doesn't work at the moment:

I didnt get the logic in previous implementation.